software_scale_context.go

```diff
@@ -0,0 +1,171 @@
+package astiav
+
+//#cgo pkg-config: libswscale
+//#include <libswscale/swscale.h>
+import "C"
+import (
+	"fmt"
+)
+
+// https://github.com/FFmpeg/FFmpeg/blob/n5.0/libswscale/swscale_internal.h#L300
+type SoftwareScaleContext struct {
+	c *C.struct_SwsContext
+	// We need to store attributes in GO since C attributes are internal and therefore not accessible
+	dstFormat C.enum_AVPixelFormat
+	dstH      C.int
+	dstW      C.int
+	flags     SoftwareScaleContextFlags
+	srcFormat C.enum_AVPixelFormat
+	srcH      C.int
+	srcW      C.int
+}
+
+func NewSoftwareScaleContext(srcW, srcH int, srcFormat PixelFormat, dstW, dstH int, dstFormat PixelFormat, flags SoftwareScaleContextFlags) *SoftwareScaleContext {
+	ssc := SoftwareScaleContext{
+		dstFormat: C.enum_AVPixelFormat(dstFormat),
+		dstH:      C.int(dstH),
+		dstW:      C.int(dstW),
+		flags:     flags,
+		srcFormat: C.enum_AVPixelFormat(srcFormat),
+		srcH:      C.int(srcH),
+		srcW:      C.int(srcW),
+	}
+
+	ssc.c = C.sws_getContext(
+		ssc.srcW,
+		ssc.srcH,
+		ssc.srcFormat,
+		ssc.dstW,
+		ssc.dstH,
+		ssc.dstFormat,
+		C.int(ssc.flags),
+		nil, nil, nil,
+	)
+	if ssc.c == nil {
+		return nil
+	}
+	return &ssc
+}
+
+func (ssc *SoftwareScaleContext) ScaleFrame(src, dst *Frame) (height int) {
+	height = int(
+		C.sws_scale(
+			ssc.c,
+			&src.c.data[0],
+			&src.c.linesize[0],
+			0,
+			C.int(src.Height()),
+			&dst.c.data[0], &dst.c.linesize[0]))
+	return
+}
+
+func (ssc *SoftwareScaleContext) updateContext() error {
+	ssc.c = C.sws_getCachedContext(
+		ssc.c,
+		ssc.srcW,
+		ssc.srcH,
+		ssc.srcFormat,
+		ssc.dstW,
+		ssc.dstH,
+		ssc.dstFormat,
+		C.int(ssc.flags),
+		nil, nil, nil,
+	)
+	if ssc.c == nil {
+		return fmt.Errorf("failed to update sws context")
+	}
+	return nil
+}
+
+func (ssc *SoftwareScaleContext) PrepareDestinationFrameForScaling(dstFrame *Frame) error {
```

Comment:
Ok libav sws is pain in ass and i Had thinking If you could help devs they have more fun with this lib :)

Reply:
If we use `sws_scale_frame` we shouldn't need either this function or to update the destination frame in GO since `sws_scale_frame` seems to be handling that automatically